### ISSUE TITLE
transfer_custom: Assertion fails when only sending data

### ISSUE THREAD

**Issue by marcfehling** [cburstedde/p4est#48]
The integration of the new p4est features in deal.II is progressing well. However, I stumbled over some unintuitive behavior of p4est while trying to transfer data using the `transfer_custom` function.

Considering the case in which we only want to send data only, and not receive any (because repartitioning has taken away any responsibility for former quadrants on a processor), an assertion in my implementation is triggered:

```
src_sizes.size(): 4
dest_sizes.size(): 0
dest_data.size(): 0
Abort: Assertion 'dest_sizes != NULL'
Abort: path/to/p4est-2.0/src/p4est_communication.c:1210
Abort
```
In deal.II, we are using containers of type `std::vector<char>` and `std::vector<int>` to provide data and size buffers with dynamically determined sizes. We access their underlying consecutive array structure with the `data` member function to provide the proper interface to `p4est`. The previous transfer of the data sizes via `transfer_fixed` (as presented in `p4est/test/test_partition2.c`) worked without any errors.

Is this a bug or intended behavior, in which I have to treat this particular case differently? I don't have a small testcase at hand right now, but I can provide one if you need more information.

@bangerth -- FYI

**Comment by marcfehling** [comment]
I just noticed that the contrary situation could occur as well, where we just want to receive data, but don't send any.

In my current implementation, I checked with an conditional if `src_data` is empty, and if so, I skip the whole `transfer_custom` function. I should not do this to consider this scenario. I'll check if the above error is still present if I remove the conditional.

**Comment by marcfehling** [comment]
Removing the conditional reveals that the case where we are only receiving data will trigger a similar assertion:
```
src_sizes.size(): 0
  content: 16 4 8 12 16
dest_sizes.size(): 5
src_data.size(): 0
dest_data.size(): 56
Abort: Assertion 'src_sizes != NULL'
Abort: path/to/p4est-2.0/src/p4est_communication.c:1211
Abort 
```
I'll change the title of this issue to be more specific.

**Comment by cburstedde** [comment]
> The integration of the new p4est features in deal.II is progressing well. However, I stumbled over some unintuitive behavior of p4est while trying to transfer data using the `transfer_custom` function.
> 
> Considering the case in which we only want to send data only, and not receive any (because repartitioning has taken away any responsibility for former quadrants on a processor), an assertion in my implementation is triggered:
> 
> ```
> src_sizes.size(): 4
> dest_sizes.size(): 0
> dest_data.size(): 0
> Abort: Assertion 'dest_sizes != NULL'
> Abort: path/to/p4est-2.0/src/p4est_communication.c:1210
> Abort
> ```
> In deal.II, we are using containers of type `std::vector<char>` and `std::vector<int>` to provide data and size buffers with dynamically determined sizes. We access their underlying consecutive array structure with the `data` member function to provide the proper interface to `p4est`. The previous transfer of the data sizes via `transfer_fixed` (as presented in `p4est/test/test_partition2.c`) worked without any errors.
> 
> Is this a bug or intended behavior, in which I have to treat this particular case differently? I don't have a small testcase at hand right now, but I can provide one if you need more information.

This is not a bug.

transfer_fixed noops if data_size == 0.

transfer_custom skips any message on the inside that has length zero, so the
nooping philosophy is the same.  While one or more or all quadrant data sizes
may be zero, the array of sizes must exist.  It is required that the data size
sent for a quadrant is the same as the data size received for this quadrant.
It may move to a different process but it's the same quadrant in the mesh.

**Comment by marcfehling** [comment]
> It is required that the data size sent for a quadrant is the same as the data size received for this quadrant. It may move to a different process but it's the same quadrant in the mesh.

I agree that the data size for any quadrant should stay the same during the transfer process. The output snippet I posted belongs to just one of many processors. Sorry for not clarifying that.

We prepared the `src_sizes` array to be of length `local_num_quadrants`. Then we apply repartitioning and we prepared `dest_sizes` to be of length `local_num_quadrants`. If for some reason one of the processors loses all ownership of quadrants (or gains ownership while not having any before), i.e. `local_num_quadrants` is zero at some point, I do not know how to handle that.

Shall I communicate the maximum size any processor has amongst all processors, and prepare the size buffers accordingly?

**Comment by bangerth** [comment]
I think what @marcfehling is saying is that he's calling a p4est function for receiving data. He knows that he won't get any data because that processor owns no quadrants, so passes in a `NULL` pointer. But p4est triggers an assertion that the pointer must be non-`NULL` even though it wasn't actually going to write into the array (because the processor owns no quadrants). In other words, the assertion is over-eager.

@marcfehling and I discussed simply allocating a one-element array and passing the pointer to it to the p4est function, even though we know that no data will be written into it. He's tried this out and it seems to work.

**Comment by cburstedde** [comment]
Agreed, you have a point. Please check out fix-transfer-assert. Problems resolved?

**Comment by marcfehling** [comment]
I just built a release tarball of that branch and tested it. `p4est-2.0.65-dfeb` works without our workaround. Thank you!

**Comment by bangerth** [comment]
@cburstedde -- let us know once this is on a release branch so that we can guard our workaround by an appropriate `#if P4EST_VERSION_GTE`

**Comment by bangerth** [comment]
Actually, we just took 2.0.65.

**Comment by cburstedde** [comment]
> Actually, we just took 2.0.65.

This can work.

Alternatively, I've added the #define P4EST_COMM_TRANSFER_NULL that you can
#ifdef for the feature.  Such feature defines will be documented in NEWS from
now on.

If I don't hear any complaints I'll merge it into develop officially.
